# Домашнее задание 1
## Анализ данных в Python (Социология)
### Темы: типы данных, условный оператор, циклы и Pandas

К каждой задаче напишите краткое объяснение того, то вы сделали, это будет влиять на оценку и таким образом показывать, что вы понимаете, что написали. Это может быть маленький комметарий в коде или пошаговое описание, можно написать в маркдауне. Если при проверке будет намек на то, что домашнее задание/часть заданий за вас сделала нейросеть, то у проверяющего есть возможность занизить оценку, это нужно в первую очередь для вас и вашего понимания курса. Не имеет смысла просто копировать ответ GPT не поняв как самому решать. Если используете нейросеть постарайтесь разобраться, с подсказкой написать код самостоятельно, а затем объяснить его. 

Всего можно набрать 10 баллов. 

**Дедлайн:** до 09:00 (утра) 24.11.2025

**Формат сдачи:** файл ipynb, название файла должно быть в виде: Фамилия_БСЦ_Группа_HW1.ipynb. За неправильно названный файл полагается штраф. Сдача ДЗ ассистенту в личку Telegram: @snovasash 


# Часть 1. Типы данных

## 1. Средний балл (списки) - 0,5 балла (по 0.1 за каждый пункт)

Дан список `student = ['Иван Питонов', 2001, [8, 7, 7, 9, 6], True]`. Выполните пункты ниже. Все выводы должны быть в формате "Средний балл студента: {средний балл}"

1. Выведите фамилию и имя Ивана в формате "Студент: {Фамилия}, {Имя}"
2. Выведите возраст Ивана, если сейчас 2020 год.
3. Напечатайте оценки Ивана через запятую.
4. Найдите средний балл Ивана, если оценки хранятся в списке внутри. Сумму элементов списка можно найти с помощью функции sum(). Выведите средний балл.
5. Если средняя оценка Ивана больше или равна 8, то он получает повышенную стипендию. Выведите True или False.

In [1]:
student = ['Иван Питонов', 2001, [8, 7, 7, 9, 6], True]

# 1. Выведите фамилию и имя Ивана в формате "Студент: {Фамилия}, {Имя}"
# Сначала обратимся к 0-му элементу списка, потом превратим строку в список и достанем нужные элементы
print(f"Студент: {student[0].split()[1]}, {student[0].split()[0]}")

# 2. Выведите возраст Ивана, если сейчас 2020 год.
# из 2020 вычтем 1-ый элемент списка
print("Возраст Ивана, если сейчас 2020 год:", 2020 - student[1])

# 3. Напечатайте оценки Ивана через запятую.
# Обратимся к 2-ому жлементу списка, для использования метода .join преобразуем список int-ов в список string-ов
print("Оценки Ивана:", ", ".join(map(str, student[2])))

# 4. Найдите средний балл Ивана, если оценки хранятся в списке внутри. Сумму элементов списка можно найти с помощью функции sum(). Выведите средний балл.
# Обратимся к 2-ому жлементу списка, просуммируем значения списка и разделим на длину списка
avg_mark = sum(student[2]) / len(student[2])
print("Средний балл студента:", avg_mark)

# 5. Если средняя оценка Ивана больше или равна 8, то он получает повышенную стипендию. Выведите True или False.
print("Студент получает повышенную стипендию:", avg_mark >= 8)



Студент: Питонов, Иван
Возраст Ивана, если сейчас 2020 год: 19
Оценки Ивана: 8, 7, 7, 9, 6
Средний балл студента: 7.4
Студент получает повышенную стипендию: False


## 2. Вложенные словари (0.25 балла)

Вам дан словарь, содержащий в себе списки словарей. Напишите код, который выведет на экран 'разработчик' из списка ключа Windows в в первом словаре из списка 2.

In [2]:
dict_of_lists = {'Список1': [{'Python': 'язык программирования'}, {'R':'язык программирования', 'LaTEX' : 'язык верстки' }],
                 'Список2' : [{'Windows' : ['операционная система', 'разработчик'], 'UNIX' : 'операционная система'},
                              {'IBM': ['компания-производитель', 'разработчик'], 'IPv6' : 'интернет-протокол' }]}


# Для выполнения задания будем идти логически с конца задания:
# который выведет на экран 'разработчик' из списка ключа Windows в в первом словаре из списка 2.
print(dict_of_lists['Список2'][0]['Windows'][1])

разработчик


## 3. Станция метро (0.25 балла)

Дан список станций желтой ветки московского метро (упорядочены как на карте):

`line = ["Третьяковская", "Марксистская", "Площадь Ильича", "Авиамоторная", "Шоссе Энтузиастов", "Перово", "Новогиреево", "Новокосино"]`

Давайте считать, что движение происходит от Третьяковской до Новокосино.

Напишите программу, которая запрашивает у пользователя название текущей станции метро (без лишнего пробела на конце) и выводит на экран сообщение вида

    Следующая станция: станция.

Считайте, что пользователь будет запрашивать станцию следующую после Новокосино, то будет выводиться сообщение: "Конечная, покиньте вагон"

Пример:

Входные данные:

    Введите текущую станцию: Площадь Ильича

Выходные данные:

    Следующая станция: Авиамоторная

Подсказка:

Чтобы вернуть индекс определенного элемента в списке, нужно воспользоваться методом .index().


In [3]:
line = ["Третьяковская", "Марксистская", "Площадь Ильича", "Авиамоторная", "Шоссе Энтузиастов", "Перово", "Новогиреево", "Новокосино"]

# получим станцию от пользователя
user_input = input("Введите текущую станцию: ").strip()

# проверим, не является ли введеная станция Новокосино, иначе вернём следующую станцию
if user_input == 'Новокосино':
    print('Конечная, покиньте вагон')
else:
    print(f"Следующая станция: {line[line.index(user_input) + 1]}")

Следующая станция: Авиамоторная


## 4. Срезы списков (0.5 балла)

Используя список `dwarves`, выведите на экран с помощью срезов следующие списки:

    ["Двалин", "Бифур", "Бофур"]
    ["Балин", "Бифур", "Бомбур", "Глоин", "Нори", "Фили", "Торин"]
    ["Оин", "Дори", "Ори", "Кили"]
    ["Торин", "Фили", "Нори", "Глоин"]
    ["Торин"]

In [4]:
dwarves = ["Балин", "Двалин", "Бифур", "Бофур", 
         "Бомбур", "Оин", "Глоин", "Дори", "Нори",
         "Ори", "Фили", "Кили", "Торин"]

# воспользуемся индексами по списку с учетом работы срезов
print(dwarves[1:4])
print(dwarves[0:13])
print(dwarves[5:6] + dwarves[7:8] + dwarves[9:10] + dwarves[11:12])
print(dwarves[12:13] + dwarves[10:11] + dwarves[8:9] + dwarves[6:7])
print(dwarves[12:13])

['Двалин', 'Бифур', 'Бофур']
['Балин', 'Двалин', 'Бифур', 'Бофур', 'Бомбур', 'Оин', 'Глоин', 'Дори', 'Нори', 'Ори', 'Фили', 'Кили', 'Торин']
['Оин', 'Дори', 'Ори', 'Кили']
['Торин', 'Фили', 'Нори', 'Глоин']
['Торин']


# Часть 2. Условный оператор и циклы

## 5. Развернуть число - 0.5 балл

Вам на вход подается произвольное целое число. "Разверните" его, не используя преобразование к строке.

Hint: нужно использовать операторы деления для получения остатка и целого числа, что мы с вами проходили.

Пример:
    
    Вход: 12345890
    Выход: 9854321

In [6]:
user_input = int(input())

print('Вход:', user_input)

print('Выход:', end='')

while user_input:
    digit = user_input % 10
    if digit:
        print(digit, end='')
    user_input = user_input // 10

Вход: 12345890
Выход:9854321

## 6. Азбука Морзе - (0.75 балла)
 
Напишите программу, которая по введённому английскому тексту кодирует каждое слово в [азбуке Морзе](https://en.wikipedia.org/wiki/Morse_code).

Каждая буква заменяется на последовательность точек и тире. В качестве тире используйте обычный дефис `-`, а в качестве точки — точку `.`.  

Например, буква **B** превращается в `-...`. Между закодированными буквами ставится ровно один пробел.  Например, слово Help превратится в `.... . .-.. .--..`. Обратите внимание, что строчные и заглавные буквы кодируются одинаково.

**Формат ввода**
Весь текст записан в единственной строке. Текст состоит из английских букв и пробелов, других символов в тексте нет. В тексте не может быть двух или более пробелов подряд. 

**Формат вывода**
Выведите каждое слово исходного текста, закодированное азбукой Морзе, на отдельной строке. Количество строк в ответе должно совпадать с количеством слов в исходном тексте.

**Пример:**

Ввод: 

Help me SOS

Вывод:

.... . .-.. .--.  
-- .  
... --- ...  

In [7]:
# во-первых, оздадим словарь соответсвий английских букв и азбуки Морзе
morse_dict = {
    'A': '.-', 'B': '-...', 'C': '-.-.',
    'D': '-..', 'E': '.', 'F': '..-.',
    'G': '--.', 'H': '....', 'I': '..',
    'J': '.---', 'K': '-.-', 'L': '.-..',
    'M': '--', 'N': '-.', 'O': '---',
    'P': '.--.', 'Q': '--.-', 'R': '.-.',
    'S': '...', 'T': '-', 'U': '..-',
    'V': '...-', 'W': '.--', 'X': '-..-',
    'Y': '-.--', 'Z': '--..'
}

# Введенный пользователем текст приведём к верхнему регистру и превратим в список
user_input_list = input().strip().upper().split()

# Чтобы вывести каждое слово отдельно на новой строке, пройдемся в цикле по полученному ранее списку
# внешним циклом мы будем итерироваться по отдельным словам,
# а циклом, который создаст генератор мы будем итерироваться по отдельным буквам каждого слова
for word in user_input_list:
    print(" ".join(morse_dict[sym] for sym in word))


.... . .-.. .--.
-- .
... --- ...


## 7. Классный прямоугольник (0.5 балла)

Разработайте класс `Rectangle`. При инициализации класс принимает два кортежа рациональных координат противоположных углов прямоугольника (со сторонами, параллельными осям координат).  

Класс должен реализовывать методы:
- `perimeter` — возвращает периметр прямоугольника;
- `area` — возвращает площадь прямоугольника.

Все результаты вычислений нужно округлить до сотых.

### Примечание

Ваше решение должно содержать только классы и функции.  

---

### Пример

**Ввод**

rect = Rectangle((3.2, -4.3), (7.52, 3.14))
print(rect.perimeter())

**Вывод**

23.52

**Ввод**

rect = Rectangle((7.52, -4.3), (3.2, 3.14))
print(rect.area())

**Вывод**

32.14


In [8]:
class Rectangle:
    def __init__(self, tup1, tup2):
        self.tup1 = tup1
        self.tup2 = tup2

    def perimeter(self):
        # метод возвращает периметр
        weight = abs(self.tup2[0] - self.tup1[0])
        height = abs(self.tup2[1] - self.tup1[1])
        # округляем до сотых
        return round(2 * (weight + height), 2)

    def area(self):
        # метод возвращает площадь
        weight = abs(self.tup2[0] - self.tup1[0])
        height = abs(self.tup2[1] - self.tup1[1])
        # округляем до сотых
        return round(weight * height, 2)
    
# протестируем работу методов с входными данными из условия задачи
rect = Rectangle((3.2, -4.3), (7.52, 3.14))
print(rect.perimeter())

assert rect.perimeter() == 23.52


rect = Rectangle((7.52, -4.3), (3.2, 3.14))
print(rect.area())

assert rect.area() == 32.14

23.52
32.14


## 8. Объединение словарей (0.75 балла)

Напишите программу, которая подсчитывает значения из произвольного списка словарей следующего вида.

Ввод:

    shops = [{'товар': 'яблоки', 'количество': 400}, {'товар': 'конфеты', 'количество': 300}, {'товар': 'яблоки', 'количество': 750}]


Вывод:

    {'яблоки': 1150, 'конфеты': 300}

In [9]:
shops = [{'товар': 'яблоки', 'количество': 400},
         {'товар': 'конфеты', 'количество': 300}, 
         {'товар': 'яблоки', 'количество': 750}]

# создаем пустой словарь для результата
total = {}

# итерируемся по каждому магазину
for shop in shops:
    product = shop['товар']
    cnt = shop['количество']
    # обновляем количество продукта
    total[product] = total.get(product, 0) + cnt

print(total)

{'яблоки': 1150, 'конфеты': 300}


# Часть 3. Pandas (5 баллов)

In [10]:
import pandas as pd
%matplotlib inline
import matplotlib.pyplot as plt

## Часть 1. Разминка 1 балл (по 0.2 за каждый пункт)

In [11]:
babies = pd.read_csv('https://raw.githubusercontent.com/pileyan/Data/master/data/babies%20names/babies_all.txt', index_col=0)

Исследуйте набор данных babies. Ответьте на вопросы.

1) Какие годы включает датасет (можно использовать unique())

2) Какое имя в датасете находится по индексом 121?

3) Cколько всего родилось детей по имени 'Aaron' за все время? Сколько в датасете девочек по имени John?

4) Насколько больше за все время родилось мальчиков чем девочек?

5) Cколько мальчиков родилось в 2010?

In [12]:
# 1) Какие годы включает датасет (можно использовать unique())
print(babies['year'].min())
print(babies['year'].max())
babies['year'].unique()

1880
2010


array([1880, 1881, 1882, 1883, 1884, 1885, 1886, 1887, 1888, 1889, 1890,
       1891, 1892, 1893, 1894, 1895, 1896, 1897, 1898, 1899, 1900, 1901,
       1902, 1903, 1904, 1905, 1906, 1907, 1908, 1909, 1910, 1911, 1912,
       1913, 1914, 1915, 1916, 1917, 1918, 1919, 1920, 1921, 1922, 1923,
       1924, 1925, 1926, 1927, 1928, 1929, 1930, 1931, 1932, 1933, 1934,
       1935, 1936, 1937, 1938, 1939, 1940, 1941, 1942, 1943, 1944, 1945,
       1946, 1947, 1948, 1949, 1950, 1951, 1952, 1953, 1954, 1955, 1956,
       1957, 1958, 1959, 1960, 1961, 1962, 1963, 1964, 1965, 1966, 1967,
       1968, 1969, 1970, 1971, 1972, 1973, 1974, 1975, 1976, 1977, 1978,
       1979, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989,
       1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000,
       2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010])

**Ответ:** мин год в выборке - 1880, а максимальный 2010. Все года которые представлены выведены выше

In [13]:
# 2) Какое имя в датасете находится по индексом 121?
babies.loc[121, 'name']

babies.iloc[121]['name']

'Ann'

**Ответ:** Ann

In [14]:
# 3) Cколько всего родилось детей по имени 'Aaron' за все время? Сколько в датасете девочек по имени John?
print("всего родилось детей по имени 'Aaron' за все время:", babies[babies['name'] == 'Aaron']['number'].sum())
print("в датасете девочек по имени John: ", babies[(babies['name'] == 'John') & (babies['sex'] == 'F')]['number'].sum())

всего родилось детей по имени 'Aaron' за все время: 527668
в датасете девочек по имени John:  21578


**Ответ:**
- всего родилось детей по имени 'Aaron' за все время: 527668
- в датасете девочек по имени John:  21578

In [15]:
# 4) Насколько больше за все время родилось мальчиков чем девочек?
print(babies[babies['sex'] == 'M']['number'].sum() - babies[babies['sex'] == 'F']['number'].sum())

2422447


**Ответ:** за все время мальчиков родилось на 2_422_447 больше, чем девочек


In [16]:
# 5) Cколько мальчиков родилось в 2010?
babies[(babies['sex'] == 'M') & (babies['year'] == 2010)]['number'].sum()

np.int64(1898382)

**Ответ:** в 2010 году родилось 1_898_382 мальчика

## Часть 2. Учимся анализировать

# Данные

Этот набор данных содержит информацию о бронировании номеров в отеле - когда было сделано бронирование, на сколько дней, количество взрослых, детей, младенцев, требуемые парковочные места и т.д.

Все персональные данные были удалены из этих данных.

Источник данных: эти данные получены из статьи "Hotel Booking Demand Datasets", написанной авторами Nuno Antonio, Ana Almeida и Luis Nunes для журнала Data in Brief, том 22, февраль 2019".

Данный находятся на гитхабе, файл: hotel_booking_data.csv

----------------------------

#### Замечание: имена, адреса email, номера телефонов и номера кредитных карт являются вымышленными. Данные о бронировании являются настоящими.

## <div style="text-align: center">Описание колонок</div>

<table><thead><tr class="rowsep-1"><th scope="col"><strong>Переменная</strong></th><th scope="col"><strong>Тип</strong></th><th scope="col"><strong>Описание</strong></th><th scope="col"><strong>Источник/Обработка данных</strong></th></tr></thead><tbody>

<tr><th scope="row"><em>ADR</em></th><td>Numeric</td><td>Средняя цена за сутки (см. <a name="bbib5" href="#bib5" class="workspace-trigger">[5]</a>)</td><td>BO, BL и TR / Вычисляется как сумма всех транзакций, делённая на количество суток (ночей)</td></tr>

<tr><th scope="row"><em>Adults</em></th><td>Integer</td><td>Количество взрослых</td><td>BO и BL</td></tr>     

<tr><th scope="row"><em>Agent</em></th><td>Categorical</td><td>Идентификатор туристического агентства, сделавшего бронирование<a name="btbl1fna" href="#tbl1fna" class="workspace-trigger"><sup>a</sup></a></td><td>BO и BL</td></tr>     

<tr><th scope="row"><em>ArrivalDateDayOfMonth</em></th><td>Integer</td><td>Дата заезда - день месяца</td><td>BO и BL</td></tr>     

<tr><th scope="row"><em>ArrivalDateMonth</em></th><td>Categorical</td><td>Дата заезда - месяц, 12 категорий: “January” - “December”</td><td>BO и BL</td></tr>     

<tr><th scope="row"><em>ArrivalDateWeekNumber</em></th><td>Integer</td><td>Дата заезда - номер недели</td><td>BO и BL</td></tr>     

<tr><th scope="row"><em>ArrivalDateYear</em></th><td>Integer</td><td>Дата заезда - год</td><td>BO и BL</td></tr>     

<tr><th scope="row"><em>AssignedRoomType</em></th><td>Categorical</td><td>Тип номера, который был сдан по факту. Иногда может отличаться от забронированного типа номера из-за операционных причин (например, забронировано больше номеров, чем есть по факту) или по запросу Клиента. Для соблюдения анонимности, вместо смыслового названия применяются коды типов номеров</td><td>BO и BL</td></tr>     

<tr><th scope="row"><em>Babies</em></th><td>Integer</td><td>Количество младенцев</td><td>BO и BL</td></tr>     

<tr><th scope="row"><em>BookingChanges</em></th><td>Integer</td><td>Количество изменений/уточнений, внесённых в бронирование с момента его создания в системе PMS до момента заезда Клиента или отмены бронирования</td><td>BO and BL/Вычисляется как кочество раз, когда менялись какие-либо атрибуты бронирования, а именно: люди, дата заезда, количество суток, тип номера или тип питания</td></tr>     

<tr><th scope="row"><em>Children</em></th><td>Integer</td><td>Количество детей</td><td>BO and BL/Количество детей - как тех, за которых было заплачено, так и гостящих бесплатно</td></tr>     

<tr><th scope="row"><em>Company</em></th><td>Categorical</td><td>Идентификатор компании, сделавшей бронирование или ответственной за оплату номера. Для соблюдения анонимности, вместо названия компании применяются идентификаторы</td><td>BO и BL.</td></tr>     

<tr><th scope="row"><em>Country</em></th><td>Categorical</td><td>Страна происхождения. Различные значения представлены в формате ISO 3155–3:2013<a name="bbib6" href="#bib6" class="workspace-trigger">[6]</a></td><td>BO, BL и NT</td></tr>     

<tr><th scope="row"><br></th><td><br></td><td><br></td><td><br></td></tr>     

<tr><th scope="row" rowspan="5"><em>CustomerType</em></th><td rowspan="5">Categorical</td><td>Тип бронирования, одна из четырёх категорий:</td><td rowspan="5">BO и BL</td></tr>     
<tr><td>Contract - по контракту;</td></tr>     
<tr><td>Group – бронирование в составе группы;</td></tr>     
<tr><td>Transient – когда бронирование не относится к контракту или группе, и не связано с другим transient-бронированием; в этой категории относятся бронирования, выполненные накануне или даже в последний момент, а также когда человек просто зашёл в отель с улицы и попросил номер</td></tr>     
<tr><td>Transient-party – когда бронирование относится к категории transient, но связано хотя бы одним другим transient-бронированием</td></tr>     

<tr><th scope="row"><em>DaysInWaitingList</em></th><td>Integer</td><td>Количество дней, в течении которых бронирование было в списке ожидания перед тем, как отправить подтверждение бронирования Клиенту</td><td>BO/Вычисляется как разница между датой подтверждения и датой создания бронирования в системе PMS</td></tr>     

<tr><th scope="row"><br></th><td><br></td><td><br></td><td><br></td></tr>     

<tr><th scope="row" rowspan="7"><em>DepositType</em></th><td rowspan="7">Categorical</td><td>Вносил ли Клиент предоплату для гарантии бронирования. Может принимать три значения:</td>
<td rowspan="2">BO and TR/Вычисляется на основе платежей в таблице transaction (TR), выполненных до даты заезда или отмены бронирования.</td></tr>     
<tr><td rowspan="3">No Deposit – предоплаты не было;</td></tr>     
<tr><td>Если платежи не было найдены, то значение “No Deposit”.</td></tr>     
<tr><td rowspan="2">Если платёж равен или превышает сумму бронирования, то значение “Non Refund”.</td></tr>     <tr><td rowspan="2">Non Refund – предоплата была равна полной сумме бронирования;</td></tr>     
<tr><td rowspan="2">В остальных случаях - значение “Refundable”</td></tr>     <tr><td>Refundable – платёж был меньше суммы бронирования.</td></tr>     

<tr><th scope="row"><em>DistributionChannel</em></th><td>Categorical</td><td>Канал распространения бронирования. Значение “TA” означает “Travel Agents”, а “TO” - "Tour Operators”</td><td>BO, BL и DC</td></tr>     

<tr><th scope="row"><em>IsCanceled</em></th><td>Categorical</td><td>Было ли бронирование отменено (1) или нет (0)</td><td>BO</td></tr>     

<tr><th scope="row"><em>IsRepeatedGuest</em></th><td>Categorical</td><td>Был ли этот Клиент в отеле раньше (1) или нет (0)</td><td>BO, BL and C/ Сверяем профиль Клиента и Клиента в бронировании. Если соответствие нашлось, и если дата создания профиля Клиента меньше даты создания бронирования в системе PMS, то считаем, что это повторный Клиент</td></tr>     

<tr><th scope="row"><em>LeadTime</em></th><td>Integer</td><td>Количество дней между датой внесения бронирования в систему PMS и датой заезда</td><td>BO and BL/ Разница между датой внесения бронирования и датой заезда</td></tr>     

<tr><th scope="row"><em>MarketSegment</em></th><td>Categorical</td><td>Сегмент рынка. Значение “TA” означает “Travel Agents”, а “TO” - “Tour Operators”</td><td>BO, BL и MS</td></tr>     

<tr><th scope="row"><br></th><td><br></td><td><br></td><td><br></td></tr>     

<tr><th scope="row" rowspan="5"><em>Meal</em></th><td rowspan="5">Categorical</td><td>Тип питания. Категории представлены стандартными для отелей пакетами питания:</td><td rowspan="5">BO, BL и ML</td></tr>     <tr><td>Undefined/SC – no meal package;</td></tr>     
<tr><td>BB – Bed &amp; Breakfast;</td></tr>     
<tr><td>HB – Half board (завтрак и ещё одно питание, обычно ужин);
</td></tr>     
<tr><td>FB – Full board (завтрак, обед и ужин)</td></tr>     

<tr><th scope="row"><em>PreviousBookingsNotCanceled</em></th><td>Integer</td><td>Количество предыдущих бронирований, которые не были отменены этим Клиентом перед текущим бронированием</td><td>BO and BL / Если профиль Клиента не был найден, то значение 0. Иначе, количество бронирований с тем же профилем, созданных перед текущим бронированием и не отменённых..</td></tr>     

<tr><th scope="row"><em>PreviousCancellations</em></th><td>Integer</td><td>Количество предыдущих бронирований, которые были отменены этим Клиентом перед текущим бронированием</td><td>BO and BL/ Если профиль Клиента не был найден, то значение 0. Иначе, количество бронирований с тем же профилем, созданных перед текущим бронированием и отменённых.</td></tr>     

<tr><th scope="row"><em>RequiredCardParkingSpaces</em></th><td>Integer</td><td>Количество парковочных мест, которые требуются Клиенту</td><td>BO и BL</td></tr>     

<tr><th scope="row"><br></th><td><br></td><td><br></td><td><br></td></tr>     

<tr><th scope="row" rowspan="4"><em>ReservationStatus</em></th><td rowspan="4">Categorical</td><td>Последний статус резервирования. Одно из трёх значений:</td>
<td rowspan="4">BO</td></tr>     <tr><td>Canceled – бронирование было отменено Клиентом;</td></tr>     
<tr><td>Check-Out – Клиент заехал в отель, но уже выехал;</td></tr>     
<tr><td>No-Show – Клиент не заехал в отель, и не сообщил почему</td></tr>     

<tr><th scope="row"><em>ReservationStatusDate</em></th><td>Date</td><td>Дата, когда было записано последнее изменение статуса. Используется вместе со значением <em>ReservationStatus</em> для того, чтобы понять, было ли бронирование отменено, или же Клиент выписался из отела</td><td>BO</td></tr>     

<tr><th scope="row"><em>ReservedRoomType</em></th><td>Categorical</td><td>Тип забронированного номера. Для соблюдения анонимности, вместо смыслового названия применяются коды типов номеров</td><td>BO и BL</td></tr>     

<tr><th scope="row"><em>StaysInWeekendNights</em></th><td>Integer</td><td>Количество суток (ночей) на выходных - суббота или воскресенье, которые Клиент забронировал или по факту гостил</td><td>BO и BL/Количество ночей в выходные из общего количества ночей </td></tr>     

<tr><th scope="row"><em>StaysInWeekNights</em></th><td>Integer</td><td>Количество суток (ночей) в будни - с понедельника по пятницу, которые Клиент забронировал или по факту гостил</td><td>BO и BL/Количество ночей в будни из общего количества ночей</td></tr>     

<tr><th scope="row"><em>TotalOfSpecialRequests</em></th><td>Integer</td><td>Количество специальных запросов, которые отправил Клиент (например, две отдельные кровати или высокий этаж)</td><td>BO и BL/Количество всех специальных запросов</td></tr></tbody></table>


# ЗАДАНИЯ (4 балла)

**Выполните задания, указанные жирным шрифтом. Ожидаемый результат приведён ниже в ячейках.**

**Пожалуйста не запускайте ячейку непосредственно над ожидаемым результатом - это приведёт к потере этого результата.**  

**Постарайтесь решить каждое из заданий с помощью одной строки кода в pandas (не каждое задание можно решить одной строкой кода, но многие можно!).**  
 
**Замечание: многие задачи можно решить разными способами!**

In [17]:
import pandas as pd

In [18]:
hotels = pd.read_csv("../data/hotel_booking_data.csv")

In [19]:
hotels.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  customer_type  \
0                        0                     0       2  ...      Transient   
1                        0                     0       2  ...      Transient   
2                        0                     1       1  ...      Transient   
3                        0                     1       1  ...      Transient   
4                        0                     2       2  ...      Transient   

    adr required_car_parking_spaces total_of_special_requests  \
0   0.0                           0                         0   
1   0.0                           0                         0   
2  75.0                           0                         0   
3  75.0                           0                         0   
4  98.0                           0                         1   

  reservation_status reservation_status_date            name  \
0          Check-Out              2015-07-01   Ernest Barnes   
1          Check-Out              2015-07-01    Andrea Baker   
2          Check-Out              2015-07-02  Rebecca Parker   
3          Check-Out              2015-07-02    Laura Murray   
4          Check-Out              2015-07-03     Linda Hines   

                         email  phone-number       credit_card  
0  Ernest.Barnes31@outlook.com  669-792-1661  ************4322  
1       Andrea_Baker94@aol.com  858-637-6955  ************9157  
2   Rebecca_Parker@comcast.net  652-885-2745  ************3734  
3            Laura_M@gmail.com  364-656-8427  ************5677  
4           LHines@verizon.com  713-226-5883  ************5498  

[5 rows x 36 columns]

In [20]:
hotels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 36 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

---
**ЗАДАНИЕ: Сколько строк содержится в этом наборе данных?**

In [21]:
# НАПИШИТЕ ВАШ КОД В ЭТОЙ ЯЧЕЙКЕ
# выведем 0-ое значение в возвращенном кортеже
hotels.shape[0]

119390

**ЗАДАНИЕ: Есть ли отсутствующие (пропущенные) данные? Если да, то в какой колонке их больше всего?**

In [22]:
# НАПИШИТЕ ВАШ КОД В ЭТОЙ ЯЧЕЙКЕ
# больше всего пропусков в company
hotels.isna().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

**ЗАДАНИЕ: Удалите колонку "company" из набора данных.**

In [23]:
# НАПИШИТЕ ВАШ КОД В ЭТОЙ ЯЧЕЙКЕ
# удалим поле company
hotels.drop(columns=['company'], inplace=True)

**ЗАДАНИЕ: Какие 5 стран встречаются в наборе данных наиболее часто?**

In [24]:
# НАПИШИТЕ ВАШ КОД В ЭТОЙ ЯЧЕЙКЕ
# посчитаем встречаемость по полю country
hotels['country'].value_counts().head()

country
PRT    48590
GBR    12129
FRA    10415
ESP     8568
DEU     7287
Name: count, dtype: int64

**ЗАДАНИЕ: Выведите имя человека, который заплатил наибольшую сумму за одни сутки (ADR - average daily rate)? Какое было значение этого ADR?**

In [25]:
# НАПИШИТЕ ВАШ КОД В ЭТОЙ ЯЧЕЙКЕ
# какое значение было этого ADR
print('Макс значение adr в выборке:', hotels['adr'].max())
# имя
print('Имя:', hotels[hotels['adr'] == hotels['adr'].max()]['name'].values[0])

Макс значение adr в выборке: 5400.0
Имя: Daniel Walter


**ЗАДАНИЕ: adr это средняя сумма за сутки для одного бронирования. Каково будет среднее значение adr для всех бронирований в этом наборе данных?**

In [26]:
# НАПИШИТЕ ВАШ КОД В ЭТОЙ ЯЧЕЙКЕ
# среднее по полю округлим до 2 знаков
print(round(hotels['adr'].mean(), 2))

101.83


**ЗАДАНИЕ: Каково среднее (mean) количество ночей, если сделать усреднение по всем бронированиям? Можете округлить до двух знаков после запятой.**

In [27]:
# НАПИШИТЕ ВАШ КОД В ЭТОЙ ЯЧЕЙКЕ
# среднее суммы полей stays_in_weekend_nights и stays_in_week_nights
print(round((hotels['stays_in_weekend_nights'] + hotels['stays_in_week_nights']).mean(), 2))

3.43


**ЗАДАНИЕ: Какова средняя общая сумма бронирования в этом наборе данных? Не *среднее за сутки*, а *среднее за бронирование*. Можете округлить до двух знаков после запятой.**

In [28]:
# НАПИШИТЕ ВАШ КОД В ЭТОЙ ЯЧЕЙКЕ
# посчитаем среднее суммы полей stays_in_weekend_nights и stays_in_week_nights
# и умножим на adr
print(
    round(
        (
            hotels['adr'] *
            (hotels['stays_in_weekend_nights'] + hotels['stays_in_week_nights'])).mean(),
        2)
)

357.85


**ЗАДАНИЕ: Выведите имена и адреса email тех, кто сделал ровно 5 специальных запросов ("Special Requests")?**

In [29]:
# НАПИШИТЕ ВАШ КОД В ЭТОЙ ЯЧЕЙКЕ
# сделаем фильтрацию и выберем необходимые поля
hotels[hotels['total_of_special_requests'] == 5][['name', 'email']]

name                          email
7860             Amanda Harper           Amanda.H66@yahoo.com
11125            Laura Sanders      Sanders_Laura@hotmail.com
14596              Tommy Ortiz            Tommy_O@hotmail.com
14921           Gilbert Miller         Miller.Gilbert@aol.com
14922           Timothy Torres         TTorres@protonmail.com
24630          Jennifer Weaver             Jennifer_W@aol.com
27288           Crystal Horton             Crystal.H@mail.com
27477           Brittney Burke       Burke_Brittney16@att.com
29906          Cynthia Cabrera    Cabrera.Cynthia@xfinity.com
29949              Sarah Floyd              Sarah_F@gmail.com
32267           Michelle Villa         Michelle.Villa@aol.com
39027           Nichole Hebert       Hebert.Nichole@gmail.com
39129         Lindsey Mckenzie       Lindsey.Mckenzie@att.com
39525           Ashley Edwards       Edwards.Ashley@yahoo.com
70114       Christopher Torres   Torres.Christopher@gmail.com
78819   Mrs. Tara Sullivan DVM           Mrs..DVM@xfinity.com
78820           Michaela Brown          MichaelaBrown@att.com
78822        Kurt Maldonado MD              KMD15@xfinity.com
97072         Jason Richardson               Jason.R@zoho.com
97099             Terri Hurley            THurley@xfinity.com
97261        Mrs. Caitlin Webb             Mrs._W@comcast.net
98410             Holly Arroyo          Arroyo_Holly@mail.com
98674          Denise Campbell             Denise_C@gmail.com
99887            Michael Smith            Michael.S42@aol.com
99888       Dr. Trevor Sellers                  Dr._S@aol.com
101569            Kayla Murphy         Kayla.Murphy@yahoo.com
102061         Taylor Martinez    Taylor.Martinez@hotmail.com
109511          Charles Wilson       Charles_Wilson@yahoo.com
109590           Tyler Allison         Tyler.A@protonmail.com
110082          Matthew Bailey         Matthew_Bailey@aol.com
110083       Charlotte Acevedo        Charlotte_A@verizon.com
111909         Darrell Brennan  Brennan_Darrell51@hotmail.com
111911          Melinda Jensen         MelindaJensen@zoho.com
113915            Terry Arnold          Arnold.Terry@zoho.com
114770             Mary Nguyen     Nguyen.Mary@protonmail.com
114909          Lindsay Cuevas      Lindsay.Cuevas40@mail.com
116455       Cynthia Hernandez   CynthiaHernandez@xfinity.com
116457          Angela Hawkins             Angela_H@gmail.com
118817              Sue Lawson            Sue.L52@comcast.net
119161         Alyssa Richards        Alyssa_Richards@aol.com

**ЗАДАНИЕ: Какой процент бронирований классифицированы как бронирования "повторными гостями"? (Для этого не следует использовать имя, воспользуйтесь колонкой is_repeated_guest)**

In [30]:
# НАПИШИТЕ ВАШ КОД В ЭТОЙ ЯЧЕЙКЕ
# посчитаем среднее, умножим на 100
print(round(100 * hotels['is_repeated_guest'].mean(), 2))

3.19


**ЗАДАНИЕ: Какие 5 фамилий встречаются в этом наборе данных наиболее часто? Можете ли Вы сделать это с помощью одной строки кода в pandas? (Для упрощения задачи можете считать фамилией самое последнее слово в колонке, например для значения "Caroline Conley MD" можете в качестве фамилии вывести MD)**

In [31]:
# НАПИШИТЕ ВАШ КОД В ЭТОЙ ЯЧЕЙКЕ
# работаем с полем name
# с помощью аксесора .str обратимся к последнему элементу списка и посчитаем встречаемость
hotels['name'].str.split().str[-1].value_counts().head()

name
Smith       2503
Johnson     1990
Williams    1618
Jones       1434
Brown       1423
Name: count, dtype: int64

**ЗАДАНИЕ: Выведите имена людей, которые бронировали номер для наибольшего количества детей и младенцев? (Отменённые бронирования тоже показывайте, то есть не важно, заехали они по факту в отель или нет - главное, что на момент бронирования номера было указано наибольшее количество детей+младенцев)**

In [32]:
# НАПИШИТЕ ВАШ КОД В ЭТОЙ ЯЧЕЙКЕ
# найдем максимум суммы количества детей и младенцев и отберем строки, в которых сумма равна максимальной
hotels[(hotels['children'] + hotels['babies']) == (hotels['children'] + hotels['babies']).max()]['name']

328        Jamie Ramirez
46619    Nicholas Parker
Name: name, dtype: object

**ЗАДАНИЕ: Какие 3 кода области (это первые три цифры телефона) встречаются в наборе данных наиболее часто?**

In [33]:
# НАПИШИТЕ ВАШ КОД В ЭТОЙ ЯЧЕЙКЕ
# с помощью аксесора str отберем три первых символа и посчитаем их встречаемость
hotels['phone-number'].str[:3].value_counts().head(3)

phone-number
799    168
185    167
541    166
Name: count, dtype: int64

**ЗАДАНИЕ: Сколько людей заселились в отель между 1м и 15м числами месяца (включая 1 и 15)? Бонусное задание: можете ли Вы сделать это с помощью одной строки кода в pandas?**

In [34]:
# НАПИШИТЕ ВАШ КОД В ЭТОЙ ЯЧЕЙКЕ
# поскольку у нас в датасете есть соответствующая колонка (arrival_date_day_of_month)
# мы отберем записи, в которых значение колонки меньше или равно 15
# далее отберем поля 'adults','children','babies' и посчитаем сумму
hotels[(hotels['arrival_date_day_of_month'] <= 15)][['adults','children','babies']].sum().sum()

np.float64(114160.0)

## Задание: **Свой собственный Github репрезиторий** (1 балл)

***

Вам необходимо создать собственный гитхаб репрезиторий, загрузить в него домашнее задание и прикрепить ссылку на него сюда. 

**Порядок действий:**
1) Создайте репризиторий гит на гитхабе
2) Склонируйте репрезиторий к себе на компьютер (это должна быть папка, где нет другого репрезитория, то есть где нет файла .git)
3) На компьютере создайте новое виртуальное окружение, в котором будете работать. Установите все нужные для работы библиотеки. 
4) Можно выполнять дз или если вы уже сделали перенесите файл с дз в папку.
4) Создайте файл .gitignore и добавьте в него виртуальное окружение (о .gitignore можно почитать в интернете, [например](https://ru.hexlet.io/courses/intro_to_git/lessons/gitignore/theory_unit))
5) Создайте файл requirements.txt, зафиксируйте библиотеки из виртуального окружения.
6) Загрузите все файлы в репрезиторий Git
7) Поздравляю с вашим первым репрезиторием! (ну может не первым...)

**В репрезитории должны быть следующие файлы:**

- Папка HW c файлом домашнего задания
- README.md с небольшой пояснительным текстом, что это за репрезиторий.
- requirements.txt (заполненный)
- лицензия MIT
- .gitignore (заполненный)

Если не очень понимаете что нужно делать, гугл вам в помощь! 

https://github.com/kamalov-artur/python-hse/tree/main